tests/utils/log_reader.py

```diff
@@ -27,101 +32,172 @@
-        self._has_data = False
-        self._lines = collections.deque(maxlen=1000)
-        self._wait_for_line_cv = threading.Condition()
-
-        self._has_new_line = False
-
-    def handle_new_line(self, line):
-        with self._lines_lock:
-            self._lines.append(line)
-        with self._wait_for_line_cv:
-            self._has_new_line = True
-            self._wait_for_line_cv.notify()
-
-    def run(self):
-        def get_line():
-            while not self._stop_event.is_set():
-                _line = self._file.readline()
-                if _line:
-                    return _line
-                time.sleep(0.1)
-
-        get_line()
-        with self._start_cv:
-            self._has_data = True
-            self._start_cv.notify()
-        self._file.seek(0)
-        while not self._stop_event.is_set():
-            line = get_line()
-            self.handle_new_line(line)
-
-    def start(self, wait_for_data=False):
-        super(LogReader, self).start()
-        if not wait_for_data:
-            return
-        with self._start_cv:
-            while not self._has_data:
-                self._start_cv.wait(0.1)
-        return
+        self._error_line_patterns = dict()
+        self._lines = list()
+
+    def add_error_check(self, pattern, message=None):
+        # type: (six.text_type, six.text_type) -> None
+        """
+        Add new regex pattern to match every incoming line.
+        In case of match, those lines are considered as errors and exception is raised.
+        :param pattern: Regular expression pattern.
+        :param message: Additional message to show when error log line is found.
+        """
+        compiled_pattern = re.compile(pattern)
+        info = ErrorPatternInfo(compiled_pattern, message)
+        self._error_line_patterns[pattern] = info
+
+    def _check_line_for_error(self, line):
+        # type: (six.text_type) -> None
+        """
+        Match new line with patters that must be caught as error.
+        """
+        for pattern, (compiled_pattern, message) in self._error_line_patterns.items():
+            if compiled_pattern.match(line):
+                if not message:
+                    message = "Log file {0} contains error line: '{1}'. Pattern: {2}.".format(
+                        self._file_path, line, pattern
+                    )
+                raise LogReaderError(message)
+
+    def _new_line_callback(self, line):
+        # type: (six.text_type) -> None
+        """
+        Callback which is invoked when new line is read.
+        """
+        self._lines.append(line)
+        self._check_line_for_error(line)
+
+    def _line_generator(self):  # type: () -> Generator
+        """Generator that reads all available lines.
+        """
+        for line in self._file:
+            line = line.strip("\n")
+            self._new_line_callback(line)
+            yield line
+
+    def _line_generator_blocking(self, timeout=10):  # type: (float) -> Generator
```

Comment:
Some tests for this ``LogReader`` class functionality itself would be nice as well :)

Reply:
There is already a file `tests/utils/tests/test_log_reader.py ` with some basic tests in it:)
Those tests are invoked by the tox target before the main tests.